system/security/plugins/testauthSecurity/testauthSecurity.cpp

```diff
@@ -0,0 +1,544 @@
+/*##############################################################################
+
+    HPCC SYSTEMS software Copyright (C) 2022 HPCC Systems®.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+
+#pragma warning( disable : 4786 )
+
+#include "basesecurity.hpp"
+
+#include "authmap.ipp"
+#include "testauthSecurity.hpp"
+
+class CUserAccess : public CSimpleInterfaceOf<IInterface>
+{
+    StringAttr userName, password;
+    SecAccessFlags defaulteECLWUScopeAccess = SecAccess_Unavailable;
+    SecAccessFlags defaultFileScopeAccess = SecAccess_Unavailable;
+    SecAccessFlags defaultFeaturAccess = SecAccess_Unavailable;
+    MapStringTo<int> featureAccesses, eclwuScopeAccesses, fileScopeAccesses;
```

Comment:
I think for fileScopes I would build a tree from all the scopes configured and save the access flag in the tree as well.  Then when I needed to authorize a scope I would parse the nodes and walk the tree until I had to authorize a node that wasn't in that part of the tree.  The last access flag found wins.  If no access flag was found  use default.

For config:
```
  defaults:
    filescope: NONE
  filescopes:
    "a": FULL
    "a::b::c": READ
    "a::x": READ
    "z::d": FULL
    "z::d::e": NONE
```
Load into a PTREE instead of a MapStringTo:
```
<root access=NONE> <!-- from default setting -->
  <a access=FULL>
      <b> 
          <c  access=READ/>
      </b>
      <x  access=READ/>
  <z>  
       <d access=FULL>
            <e access=NONE/>
       </d>
  </z>
</root>
```
Then at runtime these example authorizations would be as follows:
```
authorizeFileScope("a::b::c::more::evenmore") ==> READ
authorizeFileScope("a::b::m") ==> FULL
authorizeFileScope("z::o1::o2") ==> NONE (DEFAULT)
authorizeFileScope("z::d::e::another") ==> NONE (From Tree)
```
- Walk the tree based on the nodes of the scope path being authorized.
- Saving the last found access attribute.
- When you have a scope node that isn't in the tree at the current level, you return the last found access flag.

Reply:
Will store in PTree.